store/src/main/java/org/apache/rocketmq/store/ConsumeQueue.java

```diff
@@ -1001,4 +1002,20 @@ public void swapMap(int reserveNum, long forceSwapIntervalMs, long normalSwapInt
     public void cleanSwappedMap(long forceCleanSwapIntervalMs) {
         mappedFileQueue.cleanSwappedMap(forceCleanSwapIntervalMs);
     }
+
+    private int getEarliestOffset(int targetOffset, ByteBuffer byteBuffer, long currentStoreTime) {
+        if (targetOffset == 0) {
+            return targetOffset;
+        }
+        int preOffset = targetOffset - CQ_STORE_UNIT_SIZE;
+        byteBuffer.position(preOffset);
+        long prePhyOffset = byteBuffer.getLong();
+        int preSize = byteBuffer.getInt();
+        long preStoreTime =
+                this.messageStore.getCommitLog().pickupStoreTimestamp(prePhyOffset, preSize);
+        if (currentStoreTime == preStoreTime) {
+            targetOffset = getEarliestOffset(preOffset, byteBuffer, preStoreTime);
+        }
```

Comment:
If the batch is large, it may cause a deep recursion. Would it be better to turn this recursion into iteration?

Reply:
Okay, I'll optimize it later